src/concat_source.rs

```diff
@@ -310,6 +354,108 @@ impl StreamChunks for ConcatSource {
   }
 }
 
+#[derive(Debug, Default)]
+struct ConcatSourceOptimizer {
+  new_children: Vec<BoxSource>,
+  current_string: Option<String>,
+  current_raw_sources: CurrentRawSources,
+  strings_as_raw_sources: HashSet<BoxSource>,
+}
+
+#[derive(Debug, Default, PartialEq)]
+enum CurrentRawSources {
+  #[default]
+  None,
+  String(String),
+  Array(Vec<String>),
+}
+
+impl ConcatSourceOptimizer {
+  #[must_use]
+  fn optimize(mut self, children: &[BoxSource]) -> Vec<BoxSource> {
+    for child in children.iter() {
+      if let Some(raw_source) =
+        downcast_ref::<RawSource>(child.as_any()).filter(|s| s.is_string())
+      {
+        if let Some(current_string) = &mut self.current_string {
+          current_string.push_str(raw_source.source().as_ref());
+        } else {
+          self.current_string.replace(raw_source.source().to_string());
+        }
+      } else {
+        if let Some(current_string) = self.current_string.take() {
+          self.add_string_to_raw_sources(current_string);
+        }
+        if self.strings_as_raw_sources.contains(child) {
+          // self.add_source_to_raw_sources(child);
+        } else {
+          if self.current_raw_sources != CurrentRawSources::None {
+            self.merge_raw_sources();
+            self.current_raw_sources = CurrentRawSources::None;
+          }
+          self.new_children.push(child.clone());
+        }
+      }
+    }
+    if let Some(current_string) = self.current_string.take() {
+      self.add_string_to_raw_sources(current_string);
+    }
+    if self.current_raw_sources != CurrentRawSources::None {
+      self.merge_raw_sources();
+    }
+    self.new_children
+  }
+
+  fn add_string_to_raw_sources(&mut self, string: String) {
+    self.current_raw_sources = match mem::take(&mut self.current_raw_sources) {
+      CurrentRawSources::None => CurrentRawSources::String(string),
+      CurrentRawSources::Array(array) => {
+        let mut array = array;
+        array.push(string);
+        CurrentRawSources::Array(array)
+      }
+      CurrentRawSources::String(existing_string) => {
+        CurrentRawSources::Array(vec![existing_string, string])
+      }
+    };
+  }
+
+  // fn add_source_to_raw_sources(&mut self, source: &BoxSource) {
```

Comment:
Need to figure out what this is doing: https://github.com/webpack/webpack-sources/blob/9f98066311d53a153fdc7c633422a1d086528027/lib/ConcatSource.js#L12